src/jit/morph.cpp

```diff
@@ -4766,15 +4766,33 @@ GenTreePtr Compiler::fgMorphMultiregStructArg(GenTreePtr arg, fgArgTabEntryPtr f
 #endif
 
 #ifdef _TARGET_ARM_
-    if (fgEntryPtr->isSplit)
+    if ((fgEntryPtr->isSplit && fgEntryPtr->numSlots + fgEntryPtr->numRegs > 4) ||
+        (!fgEntryPtr->isSplit && fgEntryPtr->regNum == REG_STK))
     {
-        if (fgEntryPtr->numSlots + fgEntryPtr->numRegs > 4)
+        // If already OBJ it is set properly already.
+        if (arg->OperGet() == GT_OBJ)
         {
             return arg;
         }
-    }
-    else if (!fgEntryPtr->isHfaRegArg && fgEntryPtr->numSlots > 4)
-    {
+
+        GenTreeLclVarCommon* lclCommon =
```

Comment:
I don't understand why `arg` would be a `GT_ADDR`. If it is a struct argument being passed by value, it should presumably either be a `GT_OBJ` or a struct-typed `GT_LCL_VAR`. On Linux/x64, where the comment below seems to have come from, an arg can be a `GT_ADDR` when it is passed by reference.